sdk/evaluation/azure-ai-evaluation/azure/ai/evaluation/_evaluators/_document_retrieval/_document_retrieval.py

```diff
@@ -0,0 +1,342 @@
+# ---------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# ---------------------------------------------------------
+import math
+import operator
+from itertools import starmap
+from typing import Dict, List, TypedDict, Optional
+from azure.ai.evaluation._evaluators._common import EvaluatorBase
+from azure.ai.evaluation._exceptions import EvaluationException
+from typing_extensions import override, overload
+
+
+RetrievalGroundTruthDocument = TypedDict(
+    "RetrievalGroundTruthDocument", {"document_id": str, "query_relevance_label": int}
+)
+
+RetrievedDocument = TypedDict(
+    "RetrievedDocument", {"document_id": str, "relevance_score": float}
+)
+
+DocumentRetrievalMetrics = TypedDict(
+    "DocumentRetrievalMetrics",
+    {
+        "ndcg@3": float,
+        "ndcg@3_result": bool,
+        "xdcg@3": float,
+        "xdcg@3_result": bool,
+        "fidelity": float,
+        "fidelity_result": bool,
+        "top1_relevance": float,
+        "top1_relevance_result": bool,
+        "top3_max_relevance": float,
+        "top3_max_relevance_result": bool,
+        "holes": int,
+        "holes_result": bool,
+        "holes_ratio": float,
+        "holes_ratio_result": bool,
+        "total_retrieved_documents": int,
+        "total_groundtruth_documents": int,
+    },
+)
+
+
+class DocumentRetrievalEvaluator(EvaluatorBase):
+    """
+    Calculate document retrieval metrics, such as NDCG, XDCG, Fidelity and Top K Relevance.
+    """
+
+    def __init__(
+        self,
+        *,
+        groundtruth_label_min: int = 0,
+        groundtruth_label_max: int = 4,
+        threshold: Optional[dict] = None,
+    ):
+        super().__init__()
+        self.k = 3
```

Comment:
Can these be set by user ?

Reply:
I had considered allowing this, but I think it will make it complicated for generating the document retrieval metrics schema, since the metric names will need to contain the k-value.  For xdcg discount factor, I also thought hiding it for now in favor of using an "optimized" value recommended by the science team might be for the best.

It's certainly possible to expose them and add documentation, but I think for the first iteration it would probably be fine to hide them and see if customers request allowing to set them.